File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        ListNode i=head;
        int n=0,prevsum=0;
        while(i!=null)
        {
            i=i.next;
            n++;
        }
        ListNode k=head;
       int j;
       for(j=0;j<n/2;j++)
       {
        k=k.next;
       }
       k=reverse(k);
       ListNode p=head;
       ListNode q=k;
       while(p!=null && q!=null)
       {
        int sum=0;
        sum=p.val+q.val;
        p=p.next;
        q=q.next;
        if(sum>prevsum)
        {
        prevsum=sum;
        }
       }
       return prevsum;
    }
    public static ListNode reverse(ListNode head)
    {
        ListNode curr=head;
        ListNode prev=null;
        ListNode fwd=null;
        while(curr!=null)
        {
            fwd=curr.next;
            curr.next =prev;
            prev=curr;
            curr=fwd;
        }
        return prev;
    }
}

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.c
/**
 * Definition for singly-linked list->
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
 struct ListNode* reverse(struct ListNode* head)
 {
    struct ListNode* prev = NULL;
    struct ListNode* curr = head;
    struct ListNode* fwd;  

    while(curr != NULL)
    {
        fwd = curr->next;

        curr->next = prev;

        prev = curr;
        curr = fwd;
    }

    return prev;
 }
int pairSum(struct ListNode* head) {
    struct ListNode* temp=head;
    struct ListNode* i=head;
    int n;
    int max=-1;
    while(temp!=NULL)
    {
        temp=temp->next;
        n++;
    }
    temp=head;
     for(int k=1;k<n/2;k++)
    {
        temp=temp->next;
    }
    struct ListNode* head2=temp->next;
    temp->next=NULL;
    head2=reverse(head2);
     struct ListNode* j=head2;
     for(int k=0;k<n/2;k++)
    {
        int sum=i->val+j->val;
        if(sum>max)
        max=sum;
        i=i->next;
        j=j->next;
    }
return max;
}